DataSet feature implementation (fixes #38)

mkgerasimenko opened the pull request:
- added the method for reading from data set;
 - added the `@Source` custom annotation;
 - refactor `Objects` model;
 - refactor tests;
 - common refactoring.

mkgerasimenko committed:
Added and changed the following:
 - added the method for reading from data set;
 - added the @Source custom annotation;
 - refactor Objects model;
 - refactor tests;
 - common refactoring.

sskorol commented on src/main/java/io/github/sskorol/customannotation/Source.java:
```diff
@@ -0,0 +1,16 @@
+import java.lang.annotation.Target;
+
+/**
+ * Custom annotation for test object.
```
Describe its main purpose please.

sskorol commented on src/main/java/io/github/sskorol/dataset/ShouldSearchForLego.java:
```diff
@@ -0,0 +1,17 @@
+ * should add new Object as a new line
+ */
+@Data
+public class ShouldSearchForLego {
```
Would be more obvious if you add `Data` or `DataSet` suffix.

sskorol commented on src/main/java/io/github/sskorol/dataset/ShouldSearchForParfume.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.Data;
+
+/**
+ * A class for processing dates of Account and Parfume.
```
You probably meant `data`?

sskorol commented on src/main/java/io/github/sskorol/model/Category.java:
```diff
@@ -0,0 +1,25 @@
+    MOVIE_TV("Movie & TV Show Release Decade"),
+    TOYS_AGE_RANGE("Toys Age Range");
+
+    private final String category;
```
it might be confusing to give to your property the same name as for class...
let's say you're calling `Category.BRAND.getCategory()`... first and last parts duplicate each other...

sskorol commented on src/main/java/io/github/sskorol/model/Parfume.java:
```diff
@@ -8,10 +9,10 @@
 
-    private String sortValue;
-    private String size;
+    private String amount;
```
seems like a candidate for moving into more abstract level...
it might be also good to change its type...

sskorol commented on src/main/java/io/github/sskorol/model/SortValues.java:
```diff
@@ -0,0 +1,22 @@
+    PRICE_LOW_TO_HIGH("Price: Low to High"),
+    PRICE_HIGH_TO_LOW("Price: High to Low"),
+    DISCOUNT_LOW_TO_HIGH("Discount: Low to High"),
+    DISCOUNT("Discount: High to Low"),
```
make it consistent with a previous keys...

sskorol commented on src/main/java/io/github/sskorol/pages/LoginPage.java:
```diff
@@ -16,10 +16,11 @@
         phantomType(usernameInput, userName);
         phantomType(passwordInput, password);
         phantomClick(signInButton);
+        return new SearchPage();
```
as we discussed before, new objects creation has no effect in terms of elements locating...

mkgerasimenko replied:
We did it this way because after login we should use methods from the Search Page, not from the Login Page.

sskorol replied:
ok, then why don't you use `PageFactory`?

sskorol commented on src/main/java/io/github/sskorol/pages/ProductPage.java:
```diff
@@ -1,42 +1,46 @@
-        return new ProductPage();
+    @Step("Select the category: \"{category}\" and color: \"{value}\".")
+    public ProductPage selectColor(final Category category, final String value) {
+        click(By.xpath(format("//h4[.=\"%s\"]/following::ul[position() mod 3]//span[text()=\"%s\"]/../..",
```
what does `mod 3` mean in this context?

sskorol commented on src/main/java/io/github/sskorol/pages/ProductPage.java:
```diff
@@ -1,42 +1,46 @@
-        selectByParameters(checkboxes, condition);
-        return new ProductPage();
+    @Step("Select the category: \"{category}\" and choose: \"{value}\".")
+    public ProductPage chooseTo(final Category category, final String value) {
```
`chooseTo` sounds confusing... try to rephrase it...

sskorol commented on src/main/java/io/github/sskorol/pages/ProductPage.java:
```diff
@@ -1,42 +1,46 @@
-        return new ProductPage();
+    @Step("Select the category: \"{category}\" and choose: \"{value}\".")
+    public ProductPage chooseTo(final Category category, final String value) {
+        click(By.xpath(format("//h4[.=\"%s\"]/following::ul[position() mod 2]//span[text()=\"%s\"]",
```
seems like you have a pretty similar xpath for several methods... try to create a more generic method then...

sskorol commented on src/main/java/io/github/sskorol/utils/ReflectUtils.java:
```diff
@@ -0,0 +1,18 @@
+        throw new UnsupportedOperationException("Illegal access to private constructor.");
+    }
+
+    public static <T> T getObjectByArgs(final Class<T> entityClass, final Object... args) {
```
try to rephrase it, as technically you're creating new instance not only by args... give it a more generic name...

sskorol commented on src/test/java/io/github/sskorol/data/DataSuppliers.java:
```diff
@@ -32,4 +34,16 @@
+    @SuppressWarnings("unchecked")
+    @DataSupplier(flatMap = true)
+    public <T> T getDataSet(final Method method) {
+        return (T) ofNullable(method.getDeclaredAnnotation(Data.class).entity().getDeclaredFields())
```
try to decompose this line... `ofNullable` should be used for annotation checking, not for fields... if there's no annotation, you'll still get NPE...

sskorol commented on src/test/java/io/github/sskorol/data/DataSuppliers.java:
```diff
@@ -32,4 +34,16 @@
+    public <T> T getDataSet(final Method method) {
+        return (T) ofNullable(method.getDeclaredAnnotation(Data.class).entity().getDeclaredFields())
+                .map(fields -> StreamEx.of(fields)
+                        .flatMap(field -> Arrays.stream(readerOf(field.getType().getAnnotation(Source.class).source())
```
try to decompose it... maybe would be better to split the entire pipeline into separate methods to make it more readable...

sskorol commented on src/test/java/io/github/sskorol/testcases/AmazonTests.java:
```diff
@@ -2,175 +2,156 @@
+        open(LoginPage.class)
+                .login(data.getAccount().getUsername(), data.getAccount().getPassword())
+                .searchFor(data.getParfume().getName())
+                .selectCategory(data.getParfume().getSubCategory())
```
method assumes `category` arg, but you're passing `subCategory`... logical issue?

sskorol commented on src/test/java/io/github/sskorol/testcases/AmazonTests.java:
```diff
@@ -2,175 +2,156 @@
+                .searchFor(data.getShoes().getName())
+                .selectCategory(data.getShoes().getSubCategory())
+                .selectColor(COLOR, data.getShoes().getColor())
+                .selectDimension(SHOE_SIZE, data.getShoes().getSize())
```
`dimension` doesn't seem to be a good word for shoes' size selection...

mkgerasimenko committed:
Changed the following:
 - added more obvious descriptions and names;
 - refactor some methods;
 - common refactoring.

mkgerasimenko commented:
@sskorol, please look at the following changes.

sskorol commented on src/main/java/io/github/sskorol/pages/SearchPage.java:
```diff
@@ -14,9 +14,9 @@
+    public ProductPage searchFor(final String text) {
         type(inputSearch, text + Keys.ENTER, enabled);
-        return this;
+        return new ProductPage();
```
seems like `this` would be better to use here... check all POs, if there're similar places, which require refactoring...

sskorol commented on src/test/java/io/github/sskorol/testcases/AmazonTests.java:
```diff
@@ -2,175 +2,154 @@
-
-        at(ProductPage.class)
-                .selectCheckboxBy(lego.getAgeRange())
+                .login(data.getAccount().getUsername(), data.getAccount().getPassword())
```
why don't you pass just `Account`?

mkgerasimenko committed:
Changed the following:
 - refactor login and test methods.

mkgerasimenko commented:
@sskorol, please look at the following changes.

sskorol commented on src/test/java/io/github/sskorol/data/DataSuppliers.java:
```diff
@@ -28,8 +32,26 @@
-                        .flatMap(dataIn -> Arrays.stream(readerOf(dataIn.source())
+                        .flatMap(dataIn -> stream(readerOf(dataIn.source())
                                 .readFrom(dataIn.source(), (Class<T>) dataIn.entity()))).toArray())
+                .orElseThrow(() -> new NoClassDefFoundError("Data class not found"));
```
`NoClassDefFoundError` is more about physical class' missing.
You may want to throw smth like `IllegalStateException`.

sskorol commented on src/test/java/io/github/sskorol/data/DataSuppliers.java:
```diff
@@ -16,7 +20,7 @@
     @SuppressWarnings("unchecked")
     @DataSupplier(flatMap = true)
-    public <T> StreamEx<T> getObject(final Method method) {
+    public <T> StreamEx<T> getData(final Method method) {
```
try giving you data providers more obvious names

mkgerasimenko committed:
Changed the following:
 - refactor names of data suppliers' methods;
 - changed the NoClassDefFoundError() to the IllegalStateException.

mkgerasimenko commented:
@sskorol, please look at the following changes.
